src/storage/readpool_context.rs

```diff
@@ -0,0 +1,146 @@
+
+    // TODO: command_duration, processing_read_duration can be merged together.
+    command_duration: LocalHistogramVec,
+    processing_read_duration: LocalHistogramVec,
+    command_keyreads: LocalHistogramVec,
+    // TODO: kv_command_counter, raw_command_counter, command_pri_counter can be merged together.
+    kv_command_counter: LocalCounterVec,
+    raw_command_counter: LocalCounterVec,
+    command_pri_counter: LocalCounterVec,
+    scan_details: LocalCounterVec,
+
+    read_flow_stats: HashMap<u64, storage::FlowStatistics>,
+}
+
+impl fmt::Debug for Context {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Context").finish()
+    }
+}
+
+impl Context {
+    pub fn new(pd_sender: Option<worker::FutureScheduler<pd::PdTask>>) -> Self {
+        Context {
+            pd_sender,
+            command_duration: SCHED_HISTOGRAM_VEC.local(),
+            processing_read_duration: SCHED_PROCESSING_READ_HISTOGRAM_VEC.local(),
+            command_keyreads: KV_COMMAND_KEYREAD_HISTOGRAM_VEC.local(),
+            kv_command_counter: KV_COMMAND_COUNTER_VEC.local(),
+            raw_command_counter: RAWKV_COMMAND_COUNTER_VEC.local(),
+            command_pri_counter: SCHED_COMMANDS_PRI_COUNTER_VEC.local(),
+            scan_details: KV_COMMAND_SCAN_DETAILS.local(),
+            read_flow_stats: HashMap::default(),
+        }
+    }
+
+    #[inline]
+    pub fn collect_command_count_and_duration(
+        &mut self,
+        cmd: &str,
+        priority: readpool::Priority,
+        is_raw: bool,
+    ) -> LocalHistogramTimer {
+        if is_raw {
+            self.raw_command_counter.with_label_values(&[cmd]).inc();
+        } else {
+            self.kv_command_counter.with_label_values(&[cmd]).inc();
+        }
+        self.command_pri_counter
+            .with_label_values(&[&priority.to_string()])
+            .inc();
+        self.command_duration
+            .with_label_values(&[cmd])
+            .start_coarse_timer()
+    }
+
+    #[inline]
+    pub fn collect_processing_read_duration(&mut self, cmd: &str) -> LocalHistogramTimer {
+        self.processing_read_duration
+            .with_label_values(&[cmd])
+            .start_coarse_timer()
+    }
+
+    #[inline]
+    pub fn collect_key_reads(&mut self, cmd: &str, count: u64) {
+        self.command_keyreads
+            .with_label_values(&[cmd])
+            .observe(count as f64);
+    }
+
+    #[inline]
+    pub fn collect_scan_count(&mut self, cmd: &str, statistics: &storage::Statistics) {
+        for (cf, details) in statistics.details() {
+            for (tag, count) in details {
+                self.scan_details
+                    .with_label_values(&[cmd, cf, tag])
+                    .inc_by(count as f64)
+                    .unwrap();
+            }
+        }
+    }
+
+    #[inline]
+    pub fn collect_read_flow(&mut self, region_id: u64, statistics: &storage::Statistics) {
+        if self.pd_sender.is_none() {
```

Comment:
It's always false in production, so remove it.

Reply:
However this function will be called during testing?